wakuv2/filter_manager.go

```diff
@@ -0,0 +1,360 @@
+package wakuv2
+
+import (
+	"context"
+	"crypto/rand"
+	"errors"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/libp2p/go-libp2p/core/peer"
+
+	"github.com/status-im/status-go/wakuv2/common"
+
+	node "github.com/waku-org/go-waku/waku/v2/node"
+	"github.com/waku-org/go-waku/waku/v2/protocol"
+	"github.com/waku-org/go-waku/waku/v2/protocol/filter"
+	"github.com/waku-org/go-waku/waku/v2/protocol/relay"
+	"github.com/waku-org/go-waku/waku/v2/protocol/subscription"
+	"go.uber.org/zap"
+	"golang.org/x/exp/maps"
+)
+
+const (
+	FilterEventAdded             int = 0
+	FilterEventRemoved           int = 1
+	FilterEventPingResult        int = 2
+	FilterEventSubscribeResult   int = 3
+	FilterEventUnsubscribeResult int = 4
+	FilterEventGetStats          int = 5
+)
+
+const pingTimeout = 10 * time.Second
+
+type FilterSubs map[string]subscription.SubscriptionSet
+
+type FilterEvent struct {
+	eventType int
+	filterID  string
+	result    bool
+	peerID    peer.ID
+	tempID    string
+	sub       *subscription.SubscriptionDetails
+	ch        chan FilterSubs
+}
+
+// Methods on FilterManager maintain filter peer health
+//
+// runFilterLoop is the main event loop
+//
+// Filter Install/Uninstall events are pusehd onto eventChan
+// Subscribe, UnsubscribeWithSubscription, IsSubscriptionAlive calls
+// are invoked from goroutines and request results pushed onto eventChan
+//
+// filterSubs is the map of filter IDs to subscriptions
+
+type FilterManager struct {
+	ctx              context.Context
+	filterSubs       FilterSubs
+	eventChan        chan (FilterEvent)
```

Comment:
when are we closing this? might not be need to be closed, but worth checking

Reply:
it's also unbuffered, so if an event is pushed after the loop is done (for whatever reason), it will lock the publisher, is that something you thought about?

Reply:
Loop is done only when context is closed. So if push to the channel is performed afterwards, the goroutine will hang?

Reply:
if that' the case, we might leak memory, since we'll have a goroutine (or multiple) hanging

Reply:
Added `eventChan` buffer.